doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,279 @@
+
+So eventually the `MAX FILE SIZE` can be either an integer or a parseable string. This will be
+Postgres compatible as well. Note: Even though Postgres specifies MB not MiB, GB not GiB, the
+multiplier is 1024 not 1000. To keep things consistent we can do the same thing as well.
+
+### User Experience
+- User will need either an AWS IAM user (to use credentials) or a AWS IAM role (to use AssumeRole)
+on their end to set up an [AWS connection](../design/20231110_aws_connections.md).
+- User needs to create the AWS connection in Materialize to use for `COPY TO ... s3://...`
+ and check with `VALIDATE CONNECTION` that it's working.
+- User will also need to give the AWS user/role for Materialize the following access to the
+s3 prefixes so that we can write the data.
+  - `s3:ListBucket`: We need this permission to make sure that we are writing to an empty path.
+  - `s3:PutObject`: This is required to actually upload files to S3.
+- If they try to run a `COPY` to s3 without all the permission, then the operation should fail
+with an error. Note, running a `VALIDATE CONNECTION` does not guarantee that the specific S3
+permissions are set up correctly. These can only be validated when the actual `COPY` command
+is run.
+- If the S3 path given already has some data in it, we should fail the operation.
+- When they run the `COPY` command, to keep things simple we'll block the UI till the operation
+is complete, similar to any other query.
+- If successful, the output of the `COPY` command should return the number of rows written to S3
+similar to a [Postgres `COPY` output](https://www.postgresql.org/docs/current/sql-copy.html) of
+`COPY <row_count>`.
+- If the operation fails midway for some reason or is cancelled by the user, we should show
+a notice to the users that incomplete data could have been written to the S3 path and will
+need to be cleaned up.
+- Later we can provide an option to the user to always append a unique query ID to the path,
+so they can just run the same query again and again without needing to clean up previous data.
+
+## Implementation
+
+### Parser
+#### S3 path
+We can add a new option value type of 'Uri', which would look like
+```rust
+enum Uri {
+  // S3 path
+  S3(String),
+  ... // If we support more object stores or even local file, they can be added here
+}
+// and implement required traits so that it can be used as a valid value type
+impl TryFromValue<Value> for Uri {
+...
+}
+```
+To start with we can accept the S3 path as a string and in follow up PRs add the ability
+to eventually accept scalar expressions.
+TODO: Check if there's already some queries where we explicitly parse scalar expressions in the parser.
+
+#### `MAX FILE SIZE` Option
+Initially supporting `MAX FILE SIZE` as integer for the bytes will be pretty simple. Later we
+can introduce a `Memory` type which can take either an integer or a parseable string like `'2 GB'`,
+similar to the implementation for `Duration`.
+
```

Comment:
Are there performance benefits to selecting a certain default max file size? Snowflake has some [recommended file sizes](https://docs.snowflake.com/en/user-guide/tables-external-intro) for reading data via external tables which we can use to choose a default.

Reply:
Not much while actually writing the data. The performance benefit is on the querying data warehouse side. Usually it's faster to query large files instead of smaller files.

Reply:
Snowflake recommends 16-256MB for all other data types including csv for performance, so 256MB can be our max file size.